Codegen args

gabrielgarciagava opened the pull request.

gabrielgarciagava committed:
Upgraded to dart 3

gabrielgarciagava committed:
Added arguments to configure which parts of the code should be generated

Fields added:

generateCopy
generateCopyWith
generateCopyUsing
generateLenses

Documentation can be found in annotations.dart

gabrielgarciagava committed:
Modified data class and change class to be private

gabrielgarciagava committed:
Modified example file to include more use cases

spkersten commented on functional_data/CHANGELOG.md:
```diff
@@ -1,3 +1,12 @@
+  -  generateCopy
+  -  generateCopyWith
+  -  generateCopyUsing
+  -  generateLenses
```
It has to mention the breaking change of making the generated classes private

spkersten commented on functional_data/lib/src/annotations.dart:
```diff
@@ -1,7 +1,37 @@
+
+  /// Whether the generator should generate the copy methods.
+  ///
+  /// When set, it overrides the settings of [generateCopyWith] and [generateCopyUsing].
```
Instead of overriding, I'd prefer it to be incompatible and have asserts on the constructor.

gabrielgarciagava replied:
Same thing for `generateLenses: true` and `generateCopyWith: false`?

spkersten commented on functional_data/lib/src/annotations.dart:
```diff
@@ -1,7 +1,37 @@
+
+  /// Whether the generator should generate the copyUsing method.
+  ///
+  /// When set to `true`, a helper class will also be generated (called `ClassName$Change`).
```
It is not about the helper class, it is about the `copyUsing` method.

gabrielgarciagava replied:
Ow yes. I meant "a (helper) class that implements copyUsing functionality". I will rephrase it

spkersten commented on functional_data/CHANGELOG.md:
```diff
@@ -1,3 +1,12 @@
+## 1.1.0
```
```suggestion
## 1.0.1
```

gabrielgarciagava replied:
I consider new annotations as new features, no?

spkersten commented on functional_data_generator/CHANGELOG.md:
```diff
@@ -1,3 +1,14 @@
+  - To migrate your code change `class Foo extends $Foo` to `class Foo extends _$Foo`
+
+
+
```
Too many empty lines

gabrielgarciagava committed:
Code review changes

- The conflicting configurations are now reported to the user as an error instead of automatically handled

spkersten commented on functional_data_generator/pubspec.yaml:
```diff
@@ -5,10 +5,10 @@ description: >-
+  analyzer: '>=5.13.0 <6.0.0'
   build: ^2.0.1
   collection: ^1.15.0
   functional_data: ^1.0.0
```
To check, it is compatible with older versions that don't have the new parameters?

gabrielgarciagava replied:
Forced functional data to 1.0.0 and still works fine (of course I cannot specify the params in the example file anymore, but generation works properly generating copy methods and lenses, it would have been a problem if lenses was being default to false).

The reason is that I am using `peek` ins the generator, which accepts the annotation not existing. However I need to deal with the nullability there.
If I were using `read` instead, then it would throw an exception.
